lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1246,8 +1246,21 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
 
       bool hadTypeError = TC.typeCheckExpression(SubExpr, DC, TypeLoc(),
                                                  CTP_Unused, options);
-      if (isDiscarded && !hadTypeError)
+
+      // If a closure expression is unused, the user might have intended
+      // to write "do { ... }".
+      auto *CE = dyn_cast<ClosureExpr>(SubExpr);
```

Comment:
`isDiscarded` here means it's in `Playground` or `REPL` context or not.
In the current master, "closure as a statement" is diagnosed in Parser regardless of that.
It's considered as a syntactical error. I don't think you *need* to take it into account.